pkg/ruler/rules/objectclient/rule_store.go

```diff
@@ -74,25 +75,56 @@ func (o *RuleStore) ListAllRuleGroups(ctx context.Context) (map[string]rules.Rul
 		return nil, err
 	}
 
+	mtx := sync.Mutex{}
 	userGroupMap := map[string]rules.RuleGroupList{}
-	for _, obj := range ruleGroupObjects {
-
-		user := decomposeRuleObjectKey(obj.Key)
-		if user == "" {
-			continue
-		}
-
-		rg, err := o.getRuleGroup(ctx, obj.Key)
-		if err != nil {
-			return nil, err
-		}
+	ch := make(chan string)
+	errCh := make(chan error)
+	maxWorkers := 4
+	wg := &sync.WaitGroup{}
+
+	// We create one file per rule group. With this, we create a pool of workers that will
+	// download all rule groups in parallel. We limit the number of workers to avoid a
+	// particular user having too many rule groups rate limiting us with the object storage.
+	for i := 0; i < maxWorkers; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for key := range ch {
+				user := decomposeRuleObjectKey(key)
+				if user == "" {
+					continue
+				}
+
+				rg, err := o.getRuleGroup(ctx, key)
+				if err != nil {
+					level.Error(util.Logger).Log("msg", "failed to get rule group", "key", key, "user", user)
+					errCh <- err
```

Comment:
Given that `ListAllRuleGroups` returns first error, this gouroutine can stop after sending error to the channel.